pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,154 @@
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.12.0"
+	otrace "go.opentelemetry.io/otel/trace"
+)
+
+func initJaegerExporter(config *CollectorConfig) (sdktrace.SpanExporter, error) {
+	return jaeger.New(jaeger.WithCollectorEndpoint(jaeger.WithEndpoint(
+		config.EndpointURL,
+	)))
+}
+
+func initWriterExporter(config *WriterConfig) (sdktrace.SpanExporter, error) {
+	var opts []stdouttrace.Option
+	if config.Pretty {
+		opts = append(opts, stdouttrace.WithPrettyPrint())
+	}
+	if !config.Timestamps {
+		opts = append(opts, stdouttrace.WithoutTimestamps())
+	}
+
+	var w io.Writer
+	switch strings.ToLower(config.Destination) {
+	case "stdout":
+		w = os.Stdout
+
+	case "stderr":
+		w = os.Stderr
+
+	default:
+		return nil, errors.New("unknown writer destination")
+	}
+	opts = append(opts, stdouttrace.WithWriter(w))
+
+	return stdouttrace.New(opts...)
+}
+
+func initResource(config *ResourceConfig) (*resource.Resource, error) {
+	rsc, err := resource.New(context.Background(),
+		resource.WithContainer(),
+		resource.WithProcess(),
+		resource.WithOS(),
+		resource.WithAttributes(
+			semconv.ServiceNameKey.String(config.ServiceName),
+			semconv.ServiceVersionKey.String(config.ServiceVersion),
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fill empty values with defaults
+	return resource.Merge(resource.Default(), rsc)
+}
+
+var initOnce sync.Once
+
+// InitTelemetry initializes the telemetry package and returns the tracer provider.
+// If telemetry is not enabled it returns a noop tracer provider instead.
+func InitTelemetry(config *Config) (otrace.TracerProvider, error) {
+	if !config.Enable {
+		return otrace.NewNoopTracerProvider(), nil
+	}
+
+	var firstTime bool
+	initOnce.Do(func() { firstTime = true })
+	if !firstTime {
+		return otel.GetTracerProvider(), nil
+	}
+
+	var (
+		exp sdktrace.SpanExporter
+		err error
+	)
+	switch strings.ToLower(config.Exporter) {
+	case "jaeger":
+		exp, err = initJaegerExporter(&config.CollectorConfig)
+
+	case "writer":
+		exp, err = initWriterExporter(&config.WriterConfig)
+
+	default:
+		return nil, errors.New("unknown telemetry exporter")
+	}
+	if err != nil {
+		return nil, err
+	}
+	bsp := sdktrace.NewBatchSpanProcessor(exp)
+
+	rsc, err := initResource(&config.ResourceConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithSpanProcessor(bsp),
+		sdktrace.WithResource(rsc),
+		sdktrace.WithSampler(sdktrace.ParentBased(
+			sdktrace.TraceIDRatioBased(config.SampleProbability),
+		)),
+	)
+	cleanupFuncs = append(cleanupFuncs, tp.Shutdown)
+
+	otel.SetTracerProvider(tp)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		propagation.Baggage{},
+	))
+	return tp, nil
+}
+
+var cleanupFuncs []func(context.Context) error
```

Comment:
I think this shouldn't be global, and the caller of `InitTelemetry` should be responsible to `Shutdown` instead.

Reply:
`otel/trace.TracerProvider` interface does not feature a `Shutdown()` method, but the `otel/sdk/trace.TracerProvider` struct does, but I preferred to stick with the interface. Would returning the struct pointer be preferred in this case?

Reply:
You can also consider returning `(Trace.TraceProvider, func(context.Context) error, error)` then, and specify that the returned function is a cleanup function.